sql: gate sql features on cluster version

justinj opened the pull request:
Fixes #22843.

I couldn't find an appropriate pg error code for these so I just used
the internal error one. David says "Feature Not Supported" will report
to the reg cluster which is not ideal.

It's sufficient to test clusters which are fully upgraded to a version,
because if a cluster version is upgraded past the acceptable version for
some node in the cluster, said node will just crash.

Release note (sql change): SQL features introduced in version 2.0 cannot
be used in clusters which are not upgraded fully to 2.0.

Commit:
sql: gate sql features on cluster version

Fixes #22843.

I couldn't find an appropriate pg error code for these so I just used
the internal error one. David says "Feature Not Supported" will report
to the reg cluster which is not ideal.

It's sufficient to test clusters which are fully upgraded to a version,
because if a cluster version is upgraded past the acceptable version for
some node in the cluster, said node will just crash.

Release note (sql change): SQL features introduced in version 2.0 cannot
be used in clusters which are not upgraded fully to 2.0.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23813)
<!-- Reviewable:end -->

knz commented on pkg/sql/create_table.go:
```diff
@@ -134,6 +134,20 @@ func (n *createTableNode) startExec(params runParams) error {
 		privs = sqlbase.NewDefaultPrivilegeDescriptor()
 	}
 
+	if !params.extendedEvalCtx.ExecCfg.Settings.Version.IsMinSupported(cluster.Version2_0) {
```
I would highly recommend to move this test to sqlbase.

Right now it misses ALTER TABLE ADD COLUMN

and CREATE VIEW.

knz replied:
and maybe other things (I'm not sure about all the code paths)

nvanbenschoten commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/acceptance/version_upgrade_test.go, line 118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7_1PRAfoZ8DRV232X0:-L7_1PRAfoZ8DRV232X1:b-gkgcri) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/acceptance/version_upgrade_test.go#L118)):*
> ```Go
> }
> 
> var features = []feature{
> ```

nit: the `[]versionStep` and `[]feature` variables should either both be local or both be global.

---

*[pkg/acceptance/version_upgrade_test.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7_0m8QchjWOcQ39D8c:-L7_0m8QchjWOcQ39D8d:b-jy58xw) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/acceptance/version_upgrade_test.go#L161)):*
> ```Go
> 	_, err = db.Exec(f.query)
> 	if actualVersion.Less(minAllowedVersion) {
> 		if err == nil {
> ```

Can we make any stronger assertion about this error? If we were using a pg error code for this we could assert that, but it sounds like there are reasons why we don't want to do that. Maybe at least make this `testutils.IsError(err, "cluster version does not support")`

---

*[pkg/acceptance/version_upgrade_test.go, line 274 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7_1a_J01wNDF-h-eHB:-L7_1a_J01wNDF-h-eHC:b-yleag8) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/acceptance/version_upgrade_test.go#L274)):*
> ```Go
> 		t.Log(step.name())
> 		step.run(ctx, t, c)
> 		if s, ok := step.(clusterVersionUpgrade); ok {
> ```

Nice! I'm really happy to see this infrastructure.

---

*[pkg/sql/create_table.go, line 141 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7_2_A9hs_-hCnvG4Fr:-L7_2_A9hs_-hCnvG4Fs:bcow1m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/sql/create_table.go#L141)):*
> ```Go
> 			if d, ok := def.(*tree.ColumnTableDef); ok {
> 				if _, ok := d.Type.(*coltypes.TJSON); ok {
> 					pgerror.CodeIndeterminateDatatypeError
> ```

Does this compile?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reporting these usages to the reg cluster is not a bad thing - it would be nice to see how often people are running into this (it would suggest that we need better documentation or automation for cluster upgrades). 

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-:-L7_7QBGrXroWjt8pXQY:b-t6aanm)*
<!-- Sent from Reviewable.io -->

justinj commented:
PTAL - cc @knz

---

Review status: 0 of 20 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/acceptance/version_upgrade_test.go, line 118 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7_1PRAfoZ8DRV232X0:-L7ahgXBUHgy48c_OBRk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/acceptance/version_upgrade_test.go#L118)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: the `[]versionStep` and `[]feature` variables should either both be local or both be global.
</blockquote></details>

Done.

---

*[pkg/acceptance/version_upgrade_test.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7_0m8QchjWOcQ39D8c:-L7ahq9HkczkO2FL56fV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/acceptance/version_upgrade_test.go#L161)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we make any stronger assertion about this error? If we were using a pg error code for this we could assert that, but it sounds like there are reasons why we don't want to do that. Maybe at least make this `testutils.IsError(err, "cluster version does not support")`
</blockquote></details>

Done.

---

*[pkg/sql/create_table.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7WPed5SXbQFRVLLt0t-r1-137:-L7ahrEkLgUJ65JzlUtU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/sql/create_table.go#L137)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

and maybe other things (I'm not sure about all the code paths)
</blockquote></details>

Done.

---

*[pkg/sql/create_table.go, line 141 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7_2_A9hs_-hCnvG4Fr:-L7ahr_Bjk7YaYLg81N5:b-96ychj) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/sql/create_table.go#L141)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this compile?
</blockquote></details>

No, I accidentally added this right before pushing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-:-L7aiu1b65JyNP0oQmWG:bwvsdy0)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 20 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/acceptance/version_upgrade_test.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7_0m8QchjWOcQ39D8c:-L7bC5BAhrm4646I0rNz:bg9bnbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/acceptance/version_upgrade_test.go#L161)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Sorry, I was unclear - assert the opposite here:

```
if !testutils.IsError(err, "cluster version does not support") {
```

---

*[pkg/sql/sqlbase/structured.go, line 863 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7bCUFJg0gmdTabwHMH:-L7bCUFJg0gmdTabwHMI:by68rjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/sqlbase/structured.go#L863)):*
> ```Go
> // Validate validates that the table descriptor is well formed. Checks include
> // both single table and cross table invariants.
> func (desc *TableDescriptor) Validate(ctx context.Context, txn *client.Txn, clusterVersion *cluster.ExposedClusterVersion) error {
> ```

nit: we usually pass `st *cluster.Settings` around, not `ExposedClusterVersion` directly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 20 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/backupccl/restore.go, line 874 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7bsa1fk79FN6ciOxNF:-L7bsa1fk79FN6ciOxNG:b-c0y31d) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/ccl/backupccl/restore.go#L874)):*
> ```Go
> 
> 		for _, table := range tables {
> 			if err := table.Validate(ctx, txn, nil); err != nil {
> ```

I think this should get a settings to prevent people from restoring a backup to a cluster to an older version than it supports.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 20 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/backupccl/restore.go, line 874 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7bsa1fk79FN6ciOxNF:-L7bsz2cew-DhHhGs4JS:bcosi4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/ccl/backupccl/restore.go#L874)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I think this should get a settings to prevent people from restoring a backup to a cluster to an older version than it supports.
</blockquote></details>

Both callers of WriteTableDescs have a settings already in them, so this should be trivial to plumb.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 20 of 20 files at r2.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/backupccl/restore.go, line 874 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7bsa1fk79FN6ciOxNF:-L7cyh8lrR-8qpos1_R3:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/ccl/backupccl/restore.go#L874)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Both callers of WriteTableDescs have a settings already in them, so this should be trivial to plumb.
</blockquote></details>

:+1:

---

*[pkg/sql/create_sequence.go, line 83 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7cylhdxR1VXiiaswRd:-L7cylhdxR1VXiiaswRe:b3pt0ds) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/create_sequence.go#L83)):*
> ```Go
> 	}
> 
> 	if err = desc.ValidateTable(&params.extendedEvalCtx.ExecCfg.Settings.Version); err != nil {
> ```

Make a method on the extendedEvalCtx. Then you can write `params.EvalContext().ClusterVersionSetting()` here and below which is much more readable. But see Nathan's comment first about passing `Settings` instead.

---

*[pkg/sql/create_table.go, line 155 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czHO_w_1IE8EKiCe0:-L7czHO_w_1IE8EKiCe1:b-urtpwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/create_table.go#L155)):*
> ```Go
> 	// Only validate the table because backreferences aren't created yet.
> 	// Everything is validated below.
> 	if err := desc.ValidateTable(&params.p.extendedEvalCtx.Settings.Version); err != nil {
> ```

ditto here and below

---

*[pkg/sql/create_view.go, line 149 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czIavwpNgwJzw1WtO:-L7czIavwpNgwJzw1WtP:b-urtpwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/create_view.go#L149)):*
> ```Go
> 	}
> 
> 	if err = desc.ValidateTable(&params.extendedEvalCtx.Settings.Version); err != nil {
> ```

ditto here and below

---

*[pkg/sql/descriptor_mutation_test.go, line 88 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7cyzmlB7IdOrIPR5z3:-L7cyzmlB7IdOrIPR5z4:b4l7or4) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/descriptor_mutation_test.go#L88)):*
> ```Go
> 	}
> 	mt.tableDesc.Mutations = nil
> 	if err := mt.tableDesc.ValidateTable(&cluster.MakeTestingClusterSettings().Version); err != nil {
> ```

make a method on `cluster` to shorten this and below.

---

*[pkg/sql/drop_index.go, line 217 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7cz4BzIjVGIsWHG0ul:-L7cz4BzIjVGIsWHG0um:baje7h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/drop_index.go#L217)):*
> ```Go
> 	}
> 
> 	if err := tableDesc.Validate(ctx, p.txn, &p.extendedEvalCtx.Settings.Version); err != nil {
> ```

`p.EvalContext().Version()`

---

*[pkg/sql/grant_revoke.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czKKGd9cUM-4qyW1x:-L7czKKGd9cUM-4qyW1y:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/grant_revoke.go#L101)):*
> ```Go
> 
> 		case *sqlbase.TableDescriptor:
> 			if err := d.Validate(ctx, p.txn, &p.extendedEvalCtx.Settings.Version); err != nil {
> ```

ditto

---

*[pkg/sql/lease.go, line 170 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czQAzY0kwEsdhLTld:-L7czQAzY0kwEsdhLTle:bitt304) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/lease.go#L170)):*
> ```Go
> 		// ValidateTable instead of Validate, even though we have a txn available,
> 		// so we don't block reads waiting for this table version.
> 		if err := table.ValidateTable(&cluster.MakeTestingClusterSettings().Version); err != nil {
> ```

Why do you have testing stuff in a non-test file?

---

*[pkg/sql/lease.go, line 355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czLjJnJzKRMjUTl2U:-L7czLjJnJzKRMjUTl2V:bx9x2qk) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/lease.go#L355)):*
> ```Go
> 			log.Infof(ctx, "publish: descID=%d (%s) version=%d mtime=%s",
> 				tableDesc.ID, tableDesc.Name, tableDesc.Version, modTime.GoTime())
> 			if err := tableDesc.ValidateTable(&s.execCfg.Settings.Version); err != nil {
> ```

ditto method here and below

---

*[pkg/sql/rename_column.go, line 147 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czUZe4WLusCp-G4kS:-L7czUZe4WLusCp-G4kT:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/rename_column.go#L147)):*
> ```Go
> 
> 	descKey := sqlbase.MakeDescMetadataKey(tableDesc.GetID())
> 	if err := tableDesc.Validate(ctx, p.txn, &p.extendedEvalCtx.Settings.Version); err != nil {
> ```

ditto

---

*[pkg/sql/rename_index.go, line 86 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czVtOMcq_gO3jEmOQ:-L7czVtOMcq_gO3jEmOR:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/rename_index.go#L86)):*
> ```Go
> 	}
> 	descKey := sqlbase.MakeDescMetadataKey(tableDesc.GetID())
> 	if err := tableDesc.Validate(ctx, p.txn, &p.extendedEvalCtx.Settings.Version); err != nil {
> ```

ditto

---

*[pkg/sql/rename_table.go, line 118 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czX2sXCKayaIwbgj4:-L7czX2toSogjMGWjJWI:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/rename_table.go#L118)):*
> ```Go
> 	newTbKey := tableKey{targetDbDesc.ID, newTn.Table()}.Key()
> 
> 	if err := tableDesc.Validate(ctx, p.txn, &p.extendedEvalCtx.Settings.Version); err != nil {
> ```

ditto

---

*[pkg/sql/schema_changer.go, line 1071 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czYSg6YIMwdyXYTi7:-L7czYSh-fhTOaCLeFh_:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/schema_changer.go#L1071)):*
> ```Go
> 						table := union.Table
> 						table.MaybeUpgradeFormatVersion()
> 						if err := table.ValidateTable(&s.execCfg.Settings.Version); err != nil {
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813)*
<!-- Sent from Reviewable.io -->

knz commented:
I do want to highlight I'm much more comfortable with this approach. Well done. Also very nice tests.

justinj commented:
Review status: 2 of 21 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/acceptance/version_upgrade_test.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7_0m8QchjWOcQ39D8c:-L7eI3OUCRwQCaAJgB7y:b-99pf5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/505f40b4642884e67cff4059c8bcb11772a2aca8/pkg/acceptance/version_upgrade_test.go#L161)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Sorry, I was unclear - assert the opposite here:

```
if !testutils.IsError(err, "cluster version does not support") {
```
</blockquote></details>

Oh, of course, my bad. It would require a little more extra logic around this, since pre-2.0 *binary* we'll fail for a different reason. We could add extra checks around which binary we're running on but I think at that point it might just make more sense to just blindly check that the feature "doesn't work" for whatever reason.

---

*[pkg/ccl/backupccl/restore.go, line 874 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7bsa1fk79FN6ciOxNF:-L7eJ2rr_-ozPICr1Lua:b3mezzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/ccl/backupccl/restore.go#L874)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

:+1:
</blockquote></details>

Thanks for catching! Fixed.

---

*[pkg/sql/create_sequence.go, line 83 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7cylhdxR1VXiiaswRd:-L7eVoszDw-WoXwufba6:b-jvxh26) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/create_sequence.go#L83)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make a method on the extendedEvalCtx. Then you can write `params.EvalContext().ClusterVersionSetting()` here and below which is much more readable. But see Nathan's comment first about passing `Settings` instead.
</blockquote></details>

I'm just grabbing the field now, seemed more lightweight than introducing an accessor.

---

*[pkg/sql/descriptor_mutation_test.go, line 88 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7cyzmlB7IdOrIPR5z3:-L7eW54HlustU2IAI-Mo:b-ylqhor) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/descriptor_mutation_test.go#L88)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

make a method on `cluster` to shorten this and below.
</blockquote></details>

Introduced a constant.

---

*[pkg/sql/drop_index.go, line 217 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7cz4BzIjVGIsWHG0ul:-L7eW7T1VzoBB62WRpTV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/drop_index.go#L217)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`p.EvalContext().Version()`
</blockquote></details>

Done.

---

*[pkg/sql/grant_revoke.go, line 101 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czKKGd9cUM-4qyW1x:-L7eW7yu-XJLZeQEHfhQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/grant_revoke.go#L101)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 170 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czQAzY0kwEsdhLTld:-L7eKv0XJBz1i74JnGx7:b5h962c) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/lease.go#L170)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why do you have testing stuff in a non-test file?
</blockquote></details>

Oversight, thank you for catching!

---

*[pkg/sql/lease.go, line 355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czLjJnJzKRMjUTl2U:-L7eW8ml4XTxoPnJWEX5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/lease.go#L355)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto method here and below
</blockquote></details>

Done.

---

*[pkg/sql/rename_column.go, line 147 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czUZe4WLusCp-G4kS:-L7eW9xERR0G1xmKeZzk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/rename_column.go#L147)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/rename_index.go, line 86 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czVtOMcq_gO3jEmOQ:-L7eW9VEYuU5eXwY0s3q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/rename_index.go#L86)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/rename_table.go, line 118 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czX2sXCKayaIwbgj4:-L7eWAXQEBmMk3ai9Wvq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/rename_table.go#L118)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer.go, line 1071 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7czYSg6YIMwdyXYTi7:-L7eWB51KGT-Pmg0qBjp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/schema_changer.go#L1071)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/structured.go, line 863 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7bCUFJg0gmdTabwHMH:-L7eWBut4TP-bEn7KVvE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/05129417fcd90fa2752945860ab4b4e85d38dc31/pkg/sql/sqlbase/structured.go#L863)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: we usually pass `st *cluster.Settings` around, not `ExposedClusterVersion` directly.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813)*
<!-- Sent from Reviewable.io -->

knz commented:
Gorgeous. Thanks for the changes!
Just some nits and this is good to go.

---

Reviewed 19 of 19 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/sql/create_table.go, line 155 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7egY3SRKuNU_9t6oGe:-L7egY3SRKuNU_9t6oGf:b-bv9e6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/create_table.go#L155)):*
> ```Go
> 	// Only validate the table because backreferences aren't created yet.
> 	// Everything is validated below.
> 	if err := desc.ValidateTable(params.p.EvalContext().Settings); err != nil {
> ```

`params.EvalContext().Settings` (let's slowly eradicate the knowledge that params knows about the planner).

---

*[pkg/sql/create_table.go, line 180 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7egUfOZhlcuBJCFUe5:-L7egUfPxSZEXNUxyBVy:bt04tme) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/create_table.go#L180)):*
> ```Go
> 	}
> 
> 	if err := desc.Validate(params.ctx, params.p.txn, params.extendedEvalCtx.Settings); err != nil {
> ```

`params.EvalContext().Settings`

---

*[pkg/sql/create_table.go, line 669 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7egcS7Pc7wgb3EJzjU:-L7egcS8VGmosbzv3Rqh:blkalkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/create_table.go#L669)):*
> ```Go
> 		return err
> 	}
> 	if err := td.ValidateTable(p.extendedEvalCtx.Settings); err != nil {
> ```

`p.EvalContext().Settings`

---

*[pkg/sql/create_view.go, line 182 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7eggR4j-T_2-u7PLcg:-L7eggR6KHOHHWOJOfzh:b1cu9nv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/create_view.go#L182)):*
> ```Go
> 		params.p.notifySchemaChange(&desc, sqlbase.InvalidMutationID)
> 	}
> 	if err := desc.Validate(params.ctx, params.p.txn, params.extendedEvalCtx.Settings); err != nil {
> ```

.EvalContext()

---

*[pkg/sql/drop_index.go, line 217 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7egkL9KLa5yMf2-1Qb:-L7egkL9KLa5yMf2-1Qc:bc5h956) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/drop_index.go#L217)):*
> ```Go
> 	}
> 
> 	if err := tableDesc.Validate(ctx, p.txn, p.extendedEvalCtx.Settings); err != nil {
> ```

`p.EvalContext().Settings`

---

*[pkg/sql/grant_revoke.go, line 101 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7egmQQTOffYa7hVTKJ:-L7egmQQTOffYa7hVTKK:bc5h956) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/grant_revoke.go#L101)):*
> ```Go
> 
> 		case *sqlbase.TableDescriptor:
> 			if err := d.Validate(ctx, p.txn, p.extendedEvalCtx.Settings); err != nil {
> ```

`p.EvalContext().Settings`

---

*[pkg/sql/rename_column.go, line 147 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7egtW2IoH9YSwfPKjh:-L7egtW2IoH9YSwfPKji:bc5h956) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/rename_column.go#L147)):*
> ```Go
> 
> 	descKey := sqlbase.MakeDescMetadataKey(tableDesc.GetID())
> 	if err := tableDesc.Validate(ctx, p.txn, p.extendedEvalCtx.Settings); err != nil {
> ```

`p.EvalContext().Settings`

---

*[pkg/sql/rename_index.go, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7egv4n_50T4EhRKR9F:-L7egv4oGZi2lNCtm0qe:bc5h956) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/rename_index.go#L86)):*
> ```Go
> 	}
> 	descKey := sqlbase.MakeDescMetadataKey(tableDesc.GetID())
> 	if err := tableDesc.Validate(ctx, p.txn, p.extendedEvalCtx.Settings); err != nil {
> ```

`p.EvalContext().Settings`

---

*[pkg/sql/rename_table.go, line 118 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-L7egwhBY_EKu47ARm79:-L7egwhBY_EKu47ARm7A:bc5h956) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/rename_table.go#L118)):*
> ```Go
> 	newTbKey := tableKey{targetDbDesc.ID, newTn.Table()}.Key()
> 
> 	if err := tableDesc.Validate(ctx, p.txn, p.extendedEvalCtx.Settings); err != nil {
> ```

`p.EvalContext().Settings`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/23813#-:-L7eh1MAMivBs6K_Byub:b-rdr6e4)*
<!-- Sent from Reviewable.io -->

justinj commented:
Review status: 14 of 21 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/sql/create_table.go, line 155 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23813#-L7egY3SRKuNU_9t6oGe:-L7emNOSDPgCmuDX9Qet:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/create_table.go#L155)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`params.EvalContext().Settings` (let's slowly eradicate the knowledge that params knows about the planner).
</blockquote></details>

Done.

---

*[pkg/sql/create_table.go, line 180 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23813#-L7egUfOZhlcuBJCFUe5:-L7emNjj9IDahvJsohhg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/create_table.go#L180)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`params.EvalContext().Settings`
</blockquote></details>

Done.

---

*[pkg/sql/create_view.go, line 182 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23813#-L7eggR4j-T_2-u7PLcg:-L7emW1V5qrFcaB_X0ng:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/create_view.go#L182)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.EvalContext()
</blockquote></details>

Done.

---

*[pkg/sql/drop_index.go, line 217 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23813#-L7egkL9KLa5yMf2-1Qb:-L7emaDmB4l469ZiWL5j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/drop_index.go#L217)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`p.EvalContext().Settings`
</blockquote></details>

Done.

---

*[pkg/sql/grant_revoke.go, line 101 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23813#-L7egmQQTOffYa7hVTKJ:-L7emdcLCY2O4qOMfsFN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/grant_revoke.go#L101)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`p.EvalContext().Settings`
</blockquote></details>

Done.

---

*[pkg/sql/rename_column.go, line 147 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23813#-L7egtW2IoH9YSwfPKjh:-L7emh42DldOOEPIXCaw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/rename_column.go#L147)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`p.EvalContext().Settings`
</blockquote></details>

Done.

---

*[pkg/sql/rename_index.go, line 86 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23813#-L7egv4n_50T4EhRKR9F:-L7emksN934PRIhAQhQw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/rename_index.go#L86)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`p.EvalContext().Settings`
</blockquote></details>

Done.

---

*[pkg/sql/rename_table.go, line 118 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23813#-L7egwhBY_EKu47ARm79:-L7empD46_Rh2aaOkBTO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad2f4251276c7611985db535b9123392399e8210/pkg/sql/rename_table.go#L118)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`p.EvalContext().Settings`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/23813)*
<!-- Sent from Reviewable.io -->

justinj commented:
Thanks for the reviews everyone!
